File: GdPicture/WinForm/C#/Forms Processing - OMR/Forms Processing/FilledForm.cs
using System.Drawing;
using System.Collections.Generic;



namespace Csharp_Forms_Processing
{
	public class FilledForm
	{
		
		public int m_Image; //Image, has to be cleared by owner
		public Rectangle m_FoundAnchor; //Position of Anchor on FilledForm
		public Rectangle[] Fields; //Rectangles Surrounding the OMR fields
		int m_NumberOfFields;
		public int[] Results; //Array to hold results of OMR image
		public int[] Confidence; // Array for confidence of each field result
		
		//Constructor
		public FilledForm(int Image, int NumberOfFields, ref List<Rectangle> TemplateFields)
		{
			m_Image = Image;
			m_NumberOfFields = NumberOfFields;
			Fields = new Rectangle[NumberOfFields + 1];
			Results = new int[NumberOfFields + 1];
			Confidence = new int[NumberOfFields + 1];
			CopyFieldsFromTemplate(TemplateFields);
		}
		
		//Copies the Coordinates of the OMR Fields from the TemplateForm
		public void CopyFieldsFromTemplate(List<Rectangle> TemplateFields)
		{
			Rectangle rect = new Rectangle();
			int counter = 0;
			
			foreach (Rectangle tempLoopVar_rect in TemplateFields)
			{
				rect = tempLoopVar_rect;
				Fields[counter] = rect;
				counter++;
			}
		}
		
		//Adjust the Field Positions in the Form depending on the location the anchor was found on it
		public void AdjustFieldsAccordingToAnchor(Rectangle NewAnchorPosition, Rectangle TemplateAnchorPosition)
		{
			m_FoundAnchor = NewAnchorPosition;
			//Calculate Difference in Position between Anchor Found on Filled Form and Original Anchor On Template Form
            int Left = NewAnchorPosition.X - TemplateAnchorPosition.X;
            int Top = NewAnchorPosition.Y - TemplateAnchorPosition.Y;
			
			//Add Difference to Current Filled Form Fields Positions
			for (int i = 0; i <= m_NumberOfFields; i++)
			{
				Fields[i].X += Left;
				Fields[i].Y += Top;
			}
		}
	}
	
}

File: GdPicture/WinForm/vb.net/Forms Processing - OMR/Forms Processing/FilledForm.vb
Option Explicit On
Option Strict On

Public Class FilledForm

    Public m_Image As Integer 'Image, has to be cleared by owner
    Public m_FoundAnchor As Rectangle 'Position of Anchor on FilledForm
    Public Fields() As Rectangle 'Rectangles Surrounding the OMR fields
    Dim m_NumberOfFields As Integer
    Public Results() As Integer 'Array to hold results of OMR image
    Public Confidence() As Integer ' Array for confidence of each field result

    'Constructor
    Public Sub New(ByRef Image As Integer, ByVal NumberOfFields As Integer, ByRef TemplateFields As List(Of Rectangle))
        m_Image = Image
        m_NumberOfFields = NumberOfFields
        ReDim Fields(NumberOfFields)
        ReDim Results(NumberOfFields)
        ReDim Confidence(NumberOfFields)
        CopyFieldsFromTemplate(TemplateFields)
    End Sub

    'Copies the Coordinates of the OMR Fields from the TemplateForm
    Public Sub CopyFieldsFromTemplate(ByRef TemplateFields As List(Of Rectangle))
        Dim rect As Rectangle
        Dim counter As Integer = 0

        For Each rect In TemplateFields
            Fields(counter) = rect
            counter += 1
        Next
    End Sub

    'Adjust the Field Positions in the Form depending on the location the anchor was found on it
    Public Sub AdjustFieldsAccordingToAnchor(ByVal NewAnchorPosition As Rectangle, ByVal TemplateAnchorPosition As Rectangle)
        m_FoundAnchor = NewAnchorPosition
        'Calculate Difference in Position between Anchor Found on Filled Form and Original Anchor On Template Form
        Dim Left As Integer = NewAnchorPosition.X - TemplateAnchorPosition.X
        Dim Top As Integer = NewAnchorPosition.Y - TemplateAnchorPosition.Y

        'Add Difference to Current Filled Form Fields Positions
        For i As Integer = 0 To m_NumberOfFields
            Fields(i).X += Left
            Fields(i).Y += Top
        Next
    End Sub
End Class
